File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include <bits/stdc++.h>
using namespace std;

int visited[101][101];

int dx[4] = {0,0,-1,1};
int dy[4] = {1,-1,0,0};
int answer=0;

int board[101][101];

void bfs(int y, int x, vector<vector<int>>maps){

    queue<pair<pair<int,int>,int>>q;
    q.push({{y,x},1});
    visited[y][x]=1;
    
    while(!q.empty()){
        
        int ay = q.front().first.first;
        int ax = q.front().first.second;
        int depth = q.front().second;
        q.pop();
        
        //cout << maps.size() -1  << " " << maps[0].size()-1 <<"\n";
        if(ay == maps.size() -1 && ax == maps[0].size() -1){
            cout << depth << "\n";
            answer = depth;
            break;
        }
        
        for(int i = 0 ; i < 4 ; i ++){
            
            int ny = ay + dy[i];
            int nx = ax + dx[i];
            //cout << ny << " " << nx << " " << depth+1 << " :1: " << "\n";

            if(visited[ny][nx] == 0 && ny >= 0 && nx >= 0 && ny < 100 && nx < 100 && board[ny][nx]==1){
                visited[ny][nx] = 1;
                //cout << ny << " " << nx << " " << depth+1 << " :2: "<< "\n";
                
                q.push({{ny,nx},depth+1});
            }
        }
        
        
    }
    
}


int solution(vector<vector<int>> maps)
{
    
    for(int i = 0 ; i < maps.size(); i++){
        for(int i1 = 0; i1 < maps[0].size(); i1++){
            board[i][i1] = maps[i][i1];
        }
    }
    
    
    
        
    bfs(0,0,maps);
    
    if(answer==0){
        return -1;
    }else{
        return answer;
    }
                  
}

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;


class Solution {
    
    static int [][] dxy = {{0,1},{1,0},{-1,0},{0,-1}};
    
    public int solution(int[][] maps) {

        int answer = 0;
        int [][] visited = new int[maps.length][maps[0].length];
        bfs(maps,visited);
        
        answer = visited[maps.length-1][maps[0].length-1];
    
        if(answer==0){
            answer=-1;
        }
    

        return answer;
    }
    
    void bfs(int [][] maps, int[][] visited){
        int x= 0;
        int y =0;
        visited[y][x]=1;
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{y,x});
        
    
        while(!q.isEmpty()){
            int [] nxy = q.peek();
            int dx = nxy[1];
            int dy = nxy[0];
            q.remove();
            for(int i = 0 ; i < 4; i++){
                int nx = dx + dxy[i][0];
                int ny = dy + dxy[i][1];
                // System.out.println(ny + " " + nx );

                if(nx>=0&&ny>=0&&nx<maps[0].length&&ny<maps.length&&maps[ny][nx]==1&&visited[ny][nx]==0){
                    visited[ny][nx] = visited[dy][dx]+1;
                    q.add(new int[]{ny,nx});
                } 
            }
        }
    }
    
}